internal/state/graph/gateway_listener.go

```diff
@@ -206,7 +211,18 @@ func validateListenerHostname(listener v1beta1.Listener) []conditions.Condition
 	return nil
 }
 
-func validateListenerAllowedRouteKind(listener v1beta1.Listener) []conditions.Condition {
+func getAndValidateSupportedKinds(listener v1beta1.Listener) ([]conditions.Condition, []v1beta1.RouteGroupKind) {
```

Comment:
I wonder if it makes sense to separate these two actions. We can package the validation as a `listenerValidator`, and then set the supported kinds in the configure function. What do you think?

Reply:
We can separate out the actions, and I did consider that when I was implementing, but it means validating twice - once along with the rest of the `listenerValidator`, and then again when we are assessing what needs to be added to the supported kinds. I'm not sure there is a neat way to split it out without simply duplicating it and returning different things from each function. Splitting out actions like that would normally be a pattern I'd prefer, but I'm not convinced it makes sense in this case as the supportedKinds is dependant on the validation?

Reply:
I understand the concern about duplicating code, but I think it's important to stick with the pattern established in the configurator. Validation should be performed during the validation step. This makes the code more organized, easier to read, and easier to extend. I also think duplication is acceptable when each function has a different responsibility or purpose.

It would be nice not to change two functions when we add support for more route types. Is there a way to extract the duplicate functionality to a helper function?

Reply:
Ok, that makes sense! What I've done is created wrapper functions - one for validation, one for getting the kinds. This means the processing of the kinds is done twice, but validation is happening in the correct step. Let me know what you think!

Reply:
This works, thanks!